---
title: "EDA_samples_salmon"
author: "Susanne Bauer"
date: "25/02/2021"
output:
    html_document: 
      code_folding: hide
      fig_caption: yes
      number_sections: yes
      toc: yes
      toc_float:
        collapsed: yes
params:
  directory: "~/Documents/DataAnalysis_Ribotag/Prion_project/"
editor_options: 
  chunk_output_type: console
---

```{r Setup, include = FALSE}
knitr::opts_knit$set(root.dir = "~/Documents/DataAnalysis_Ribotag/Prion_project")
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(fig.height = 6, fig.width = 8, fig.align = 'center')
```


```{r load libraries, include= FALSE}
library(tximport)
library(umap)
library(dplyr)
library(Hmisc)
library(ggplot2)
library(ggrepel)
library(corrplot)
library(RColorBrewer)
library(factoextra)
library(pheatmap)
library(ggcorrplot)
library(viridisLite)

dir <- params$directory
```


```{r set color scheme}
group_color_scheme <- c(FFI = "#B74D67", CJD = "#1D599C", Control = "#C7B6DD")
celltype_color_scheme <- c(Gad2 = "#50C1CA", PV = "#8B6EDA", SST = "#E0D42A", vGlut2 = "#D55E00")
region_color_scheme <- c(Cerebrum = "black", Cerebellum = "#B7B6BB")
fraction_color_scheme <- c(IP = "#157F1B" , total = "#E8B94D" )

ann_colors <- list(Group = group_color_scheme,
                   Cell_type = celltype_color_scheme,
                   Region = region_color_scheme,
                   Fraction = fraction_color_scheme)
```


```{r load all samples, include=FALSE}

metadata <- read.csv(paste0(dir,"metadata/prionsamples_meta.csv"), header = T, sep = ",")
cols <- c("Age", "Prep_date","Death_day","seq_lane","batch", "Group")
metadata[cols] <- lapply(metadata[cols], factor) 
metadata$labels <- gsub(pattern = "_", replacement = " ", metadata$nfcore_id)

#list salmon output files (quant.sf) for selected samples
files <- file.path(paste0(dir, "raw_data/"),metadata$nfcore_id, "quant.sf")

#assign sample id to file names to rename samples based on id in txi output
names(files) <- metadata$sample_id  

tx2gene <- read.table(paste0(dir, "metadata/salmon_tx2gene.tsv"))

#read in salmon output files 
txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
#tximport generates several output files. "count" matrix containes raw counts summarized to gene level, which are used for DESeq2 analysis. "abundance" gives tpm values summarized to gene level.


#check that sample order in count and metadata files match
sample_ids <- metadata$sample_id
all(colnames(txi$counts) == sample_ids)

all_TPMs <- txi$abundance

#Check total sum TPM (should be 1 million, including ERCC reads!)
TPM_sum <- data.frame(colSums(all_TPMs))
which(TPM_sum == 1000000)

#extract and subset ERCC counts
ERCC_idx <- grep(pattern = "ERCC*", rownames(all_TPMs))
ERCC_TPMs <- data.frame(all_TPMs[ERCC_idx,])
all_TPMs <- data.frame(all_TPMs[-ERCC_idx,])

rownames(all_TPMs) <- stringr::str_replace(rownames(all_TPMs), pattern = ".[0-9]+$", replacement = "")

#replace ensembl id with gene symbol, keep only protein coding genes

library(biomaRt)
mouse_mart = useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
df_mm_genes_ensembl <- getBM(attributes = c("ensembl_gene_id", "entrezgene_id", "external_gene_name", "gene_biotype", "chromosome_name", "start_position", "end_position", "strand", "description"), 
             mart = mouse_mart)
detach(package:biomaRt)

ProteinCodingGenes <- df_mm_genes_ensembl %>% filter(gene_biotype == "protein_coding")

samples <- colnames(all_TPMs)
gene_names <- ProteinCodingGenes[match(rownames(all_TPMs), ProteinCodingGenes$ensembl_gene_id), c("ensembl_gene_id", "external_gene_name")]

idx_PC <- rownames(all_TPMs) %in% gene_names$ensembl_gene_id
all_TPM_prot <- all_TPMs[idx_PC,]
```



```{r plot detected genes per sample}
#plot number of detected genes per sample (TPM >1), get average
det_genes <- colSums(all_TPM_prot >1)
barplot(det_genes, main = "Number of detected genes (TPM >1)")
mean_genes <- round(mean(det_genes))
paste0("Mean number of detected genes per sample: ", mean_genes)

annot <- metadata %>% dplyr::select(sample_id, Group, Region, Fraction, Cell_type)
det_genes <- merge(det_genes, annot, by.x = 0, by.y = 1)


#boxplot of detected genes per sample
plot_det_genes <- ggplot(det_genes, aes(x = Group, y = x, fill = Cell_type))+
  geom_boxplot(width = 0.5)+
  geom_text_repel(aes(label = ifelse(x < 11500, Row.names,"")), show.legend = F)+
  scale_fill_manual(values = celltype_color_scheme)+
  theme_bw()+
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 14),
        plot.title = element_text(size = 18),
        plot.subtitle = element_text(size = 14),
        axis.ticks = element_blank(),
        plot.title.position = "plot")+
  labs(title = "Detected genes in all RiboTag IPs and total RNA samples",
       subtitle = "Number of protein-coding genes with TPM > 1",
       x = "",
       y = "Protein-coding genes",
       fill = "Cell type")+
  facet_grid(~Fraction)

plot_det_genes
ggsave(file = paste0(dir, "figures/QC/detected_genes_all_samples.png"), plot = plot_det_genes, dpi = 300, width = 18, height = 10, units = "cm")

```


#MultiQC stats
```{r}
#Mapped and Aligned samples
ggplot(metadata, aes(x = sample_id, color = Fraction))+
  geom_point(aes(y = Mapped_reads_samtools_mio), shape = 1)+
  geom_point(aes(y = Aligned_Star_mio), shape = 4)+
  geom_hline(yintercept = 20, color = "black", linetype = "dashed")+
  geom_hline(yintercept = 150, color = "black", linetype = "dashed")+
  geom_text_repel(aes(x = sample_id, y = Aligned_Star_mio, label = ifelse(Aligned_Star_mio < 20, labels, "")), show.legend = F)+
  geom_text_repel(aes(x = sample_id, y = Mapped_reads_samtools_mio, label = ifelse(Mapped_reads_samtools_mio > 150, sample_id, "")), show.legend = F)+
  scale_color_manual(values = fraction_color_scheme)+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        panel.grid.major.x = element_blank())+
  labs(title = "Reads Mapped in bam file (o) and uniquely mapped reads (x)",
       x = "Samples",
       y = "Reads (million)")

ggplot(metadata, aes(x = sample_id, color = Group))+
  geom_point(aes(y = Aligned_Star_mio), shape = 4)+
  geom_hline(yintercept = 20, color = "black", linetype = "dashed")+
  geom_text_repel(aes(x = sample_id, y = Aligned_Star_mio, label = ifelse(Aligned_Star_mio < 20 |Aligned_Star_mio > 60, labels, "")), show.legend = F)+
  scale_color_manual(values = group_color_scheme)+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        panel.grid.major.x = element_blank())+
  labs(title = "Uniquely mapped reads (x)",
       x = "Samples",
       y = "Reads (million)")+
  facet_grid(~Fraction)


#Percentage of uniqyely mapped reads
unique_reads <- ggplot(metadata, aes(x = Cell_type, y = Percent_Unique_mapped_Star, fill = Group))+
  geom_jitter(position = position_jitterdodge())+
  geom_boxplot(outlier.shape = NA)+
  geom_hline(yintercept = 80, linetype = "dashed")+
  geom_text_repel(aes(label = ifelse(round(Percent_Unique_mapped_Star) <= 80, labels, '')), show.legend = F, position = position_jitterdodge())+
  scale_fill_manual(values = group_color_scheme)+
  theme_bw()+
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        plot.title.position = "plot")+
  labs(title = "Percentage of uniquely mapped reads (STAR)",
       x = "",
       y = "Mapped reads (%)",
       fill = "Disease")+
  facet_grid(~Fraction)
unique_reads
ggsave(file = paste0(dir, "figures/QC/mapped_reads_all.png"), plot = unique_reads, dpi = 300, width = 25, height = 15, units = "cm")


#duplicates
ggplot(metadata, aes(x = Cell_type, y = Percent_duplication_picard, color = Fraction))+
  geom_jitter(width = 0.15)+
  geom_text_repel(aes(label = ifelse(Percent_duplication_picard > 90, labels, "")), show.legend = F)+
  scale_color_manual(values = fraction_color_scheme)+
  theme_bw()+
  labs(title = "Duplication of reads",
       x = "",
       y = "Percentage Duplication (Picard)")+
  facet_grid(~Group)


#Proper read pairs
ggplot(metadata, aes(x = sample_id, y = Perc_porperly_paired_samtools, color = Fraction))+
  geom_point()+
  geom_text_repel(aes(label= ifelse(Perc_porperly_paired_samtools < 99.8, labels,'')), show.legend = F)+
  scale_color_manual(values = fraction_color_scheme)+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        panel.grid.major.x = element_blank())+
  labs(title = "Percentage properly paired reads",
       subtitle = "all samples",
       x = "",
       y = "Properly paired reads (%)")


#Aligned reads pre and post exclusion of samples
ggplot(metadata, aes(x = Cell_type, y = Aligned_Star_mio, fill = Group))+
  geom_boxplot()+
  geom_text_repel(aes(label = ifelse(Aligned_Star_mio < 25, labels, "")), show.legend = F)+
  #geom_hline(yintercept = 25, color = "black")+
  #geom_text_repel(aes(label= ifelse((Aligned_Star_mio > 1.5*IQR(Aligned_Star_mio) |(Aligned_Star_mio < 1.5*IQR(Aligned_Star_mio))), sample_id,'')))+
  scale_fill_manual(values = group_color_scheme)+
  theme_bw()+
  theme(panel.grid = element_blank())+
  labs(title = "Aligned reads (STAR)",
       subtitle = "all samples",
       x = "",
       y = "Aligned reads (million)")


#Aligned reads by disease
ggplot(metadata, aes(x = sample_id, y = Aligned_Star_mio, color = Group))+
  geom_point()+
  geom_text_repel(aes(label= ifelse((Aligned_Star_mio > 1.5*IQR(Aligned_Star_mio) |(Aligned_Star_mio < 1.5*IQR(Aligned_Star_mio))), labels,'')), show.legend = F)+
  scale_color_manual(values = group_color_scheme)+
  theme_bw()+
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank())+
  labs(title = "Aligned reads (STAR)",
       subtitle = "samples excluded",
       x = "",
       y = "Aligned reads (million)")

#Percentage rRNA reads
ggplot(metadata, aes(x = sample_id, y = Percent_rRNA, color = Fraction))+
  geom_point()+
  geom_hline(yintercept = 10, color = "black")+
  geom_text_repel(aes(label = ifelse(Percent_rRNA > 10, labels, "")), show.legend = F)+
  scale_color_manual(values = fraction_color_scheme)+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        panel.grid = element_blank(),
        plot.title.position = "plot")+
  labs(title = "Percentage of reads mapping to rRNA features",
       x = "Samples",
       y = "Reads (%)")
  
```


#PCA all samples
```{r}

# PCA_by_fraction <- function(fraction, TPM_df, metadata){
#   sel_fraction <- metadata %>%
#   filter(Fraction == fraction)
# 
#   data <- TPM_df[,sel_fraction$sample_id]
#   #calculate variance across all samples (columns) for each gene (row)
#   var <- apply(data,1,var)
#   
#   #keep only top x % of variabe genes
#   perc <- round(length(var)*0.1)
#   top_var <- names(var[order(var, decreasing = T)][1:perc])
#   data <- data[top_var,]
#   
#   PCA_data <- log2(t(data) + 1)
#   PCA <- prcomp(PCA_data)
#   return(PCA)
# }
# 
# IP_PCA <- PCA_by_fraction(fraction = "IP", TPM_df = all_TPM_prot, metadata = metadata)
# 
# PCs <- data.frame(IP_PCA$x)
# 
# percentVar <- round(100 * attr(pcaData, "percentVar"))
# 
# annot <- metadata %>%
#   filter(sample_id %in% rownames(PCs))%>%
#   dplyr::select(sample_id, Cell_type, Group, Fraction, Region)
# 
# IP_PCA <- ggplot(PCs, aes(x = PC1, y = PC2))+
#   geom_point(aes(fill = annot$Cell_type, shape = annot$Group, color = annot$Region), size = 3)+
#   scale_shape_manual(values = c(21, 22, 24))+
#   scale_color_manual(values = region_color_scheme)+
#   scale_fill_manual(values = celltype_color_scheme)+
#   guides(fill = guide_legend(override.aes = list(color =  celltype_color_scheme)),
#         color = guide_legend(override.aes = list(shape = 21)))+
#   theme_bw()+
#   theme(panel.grid = element_blank(),
#         plot.title.position = "plot")+
#   labs(title = "PCA all IP Samples",
#        x = "PC 1",
#        y = "PC 2",
#        fill = "Cell type", 
#        shape = "Genotype", 
#        color = "Brain Region")
# IP_PCA
# 
# 
# #PCA for total RNA samples
# total_PCA <- PCA_by_fraction(fraction = "total", TPM_df = all_TPM_prot, metadata = metadata)
# PCs <- data.frame(total_PCA$x)
# 
# annot <- metadata %>%
#   filter(sample_id %in% rownames(PCs))%>%
#   dplyr::select(sample_id, Cell_type, Group, Fraction, Region)
# 
# total_PCA <- ggplot(PCs, aes(x = PC1, y = PC2))+
#   geom_point(aes(fill = annot$Cell_type, shape = annot$Group, color = annot$Region), size = 3)+
#   scale_shape_manual(values = c(21, 22, 24))+
#   scale_color_manual(values = region_color_scheme)+
#   scale_fill_manual(values = celltype_color_scheme)+
#   guides(fill = guide_legend(override.aes = list(color =  celltype_color_scheme)),
#          color = guide_legend(override.aes = list(shape = 21)))+
#   theme_bw()+
#   theme(panel.grid = element_blank(),
#         plot.title.position = "plot")+
#   labs(title = "PCA all total RNA samples",
#        x = "PC 1",
#        y = "PC 2",
#        fill = "Cell type", 
#        shape = "Genotype", 
#        color = "Brain Region")
# total_PCA
# ggsave(file = paste0(dir, "figures/QC/PCA_alltotal.png"), plot = total_PCA, dpi = 300)

```


#save output 
```{r save output, message = FALSE}
#save analysis details (date, included samples)
run_info <- list(date(), row.names(meta), sessionInfo())
names(run_info) <- c("Run date", "Analyzed samples", "Session Info")
writeLines(capture.output(run_info), paste("Run_info_EDA.txt"))

print(run_info)

date <- Sys.time()
save.image(file = paste0(dir,"R_sessions/EDA_", date,".RData"))
```



